util/metrics-service/src/config.rs

```diff
@@ -0,0 +1,191 @@
+use std::{collections::HashMap, net::SocketAddr, time::Duration};
+
+use metrics_core::Observe;
+use metrics_exporter_http::HttpExporter;
+use metrics_exporter_log::LogExporter;
+use metrics_observer_json::JsonBuilder;
+use metrics_observer_prometheus::PrometheusBuilder;
+use metrics_observer_yaml::YamlBuilder;
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+
+use ckb_async_runtime::Handle;
+
+/* Examples:
+ * ```toml
+ * [metrics]
+ * threads = 3
+ * histogram_window = 60
+ * histogram_granularity = 1
+ * upkeep_interval = 500
+ * [metrics.exporter.prometheus]
+ * target = { type = "http", listen_address = "127.0.0.1:8100" }
+ * format = { type = "prometheus" }
+ * [metrics.exporter.log_yaml]
+ * target = { type = "log", level = "warn", interval = 600 }
+ * format = { type = "yaml" }
+ * [metrics.exporter.log_json]
+ * target = { type = "log", level = "error", interval = 900 }
+ * format = { type = "json" }
+ * ```
+ */
+#[derive(Default, Clone, Debug, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    #[serde(default)]
+    pub threads: usize,
+
+    #[serde(default)]
+    pub histogram_window: u64, // seconds
+    #[serde(default)]
+    pub histogram_granularity: u64, // seconds
+    #[serde(default)]
+    pub upkeep_interval: u64, // milliseconds
+
+    #[serde(default)]
+    pub exporter: HashMap<String, Exporter>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Exporter {
+    pub target: Target,
+    pub format: Format,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "type")]
+pub enum Target {
+    Log {
+        level: LogLevel,
+        interval: u64, // seconds
+    },
+    Http {
+        listen_address: String,
+    },
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "type")]
+pub enum Format {
+    Json {
+        #[serde(default)]
+        pretty: bool,
+    },
+    Yaml,
+    Prometheus,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum LogLevel {
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+impl Config {
+    // .map_err(|err| format!("metrics exporter {}: {}", name, err))?;
+
+    pub fn check_exporter_name(name: &str) -> Result<(), String> {
+        if name.is_empty() {
+            return Err("the name of exporter shouldn't be empty".to_owned());
+        }
+        match Regex::new(r"^[0-9a-zA-Z_-]+$") {
+            Ok(re) => {
+                if !re.is_match(&name) {
+                    return Err(format!(
+                        "invaild exporter name \"{}\", only \"0-9a-zA-Z_-\" are allowed",
+                        name
+                    ));
+                }
+            }
+            Err(err) => {
```

Comment:
Example:
```toml
[metrics.exporter."I+can~use*any#characters^between@double%quotes"] 
```

Reply:
What you talk about is `re.is_match`. I mean `Re::new()` will not return errors in our case.

Reply:
Since `Regex::new()` return `Result` and we can handle it in this function, I think `match Error` is OK.
I think `panic` and `expect` should be only used in the context  which can't do error handling.